common/log.h

```diff
@@ -0,0 +1,658 @@
+//  Makes LOG() and LOG_TEE() produce no output,
+//  untill enabled back.
+#define LOG_DISABLE() _log_disable()
+
+// INTERNAL, DO NOT USE
+inline FILE *_log_disable()
+{
+    return _log_handler1(true, LogTriStateTrue);
+}
+
+// Enables logs at runtime.
+#define LOG_ENABLE() _log_enable()
+
+// INTERNAL, DO NOT USE
+inline FILE *_log_enable()
+{
+    return _log_handler1(true, LogTriStateFalse);
+}
+
+// Sets target fir logs, either by a file name or FILE* pointer (stdout, stderr, or any valid FILE*)
+#define LOG_SET_TARGET(target) _log_set_target(target)
+
+// INTERNAL, DO NOT USE
+inline FILE *_log_set_target(std::string filename) { return _log_handler1(true, LogTriStateSame, filename); }
+inline FILE *_log_set_target(FILE *target) { return _log_handler2(true, LogTriStateSame, target); }
+
+// INTERNAL, DO NOT USE
+inline FILE *log_handler() { return _log_handler1(); }
+
+inline void log_test()
+{
+    LOG_DISABLE();
+    LOG("01 Hello World to nobody, because logs are disabled!\n")
+    LOG_ENABLE();
+    LOG("02 Hello World to default output, which is \"%s\" ( Yaaay, arguments! )!\n", LOG_STRINGIZE(LOG_TARGET))
+    LOG_TEE("03 Hello World to **both** default output and " LOG_TEE_TARGET_STRING "!\n")
+    LOG_SET_TARGET(stderr);
+    LOG("04 Hello World to stderr!\n")
+    LOG_TEE("05 Hello World TEE with double printing to stderr prevented!\n")
+    LOG_SET_TARGET(LOG_DEFAULT_FILE_NAME);
+    LOG("06 Hello World to default log file!\n")
+    LOG_SET_TARGET(stdout);
+    LOG("07 Hello World to stdout!\n")
+    LOG_SET_TARGET(LOG_DEFAULT_FILE_NAME);
+    LOG("08 Hello World to default log file again!\n")
+    LOG_DISABLE();
+    LOG("09 Hello World _1_ into the void!\n")
+    LOG_ENABLE();
+    LOG("10 Hello World back from the void ( you should not see _1_ in the log or the output )!\n")
+    LOG_DISABLE();
+    LOG_SET_TARGET("llama.anotherlog.log");
+    LOG("11 Hello World _2_ to nobody, new target was selected but logs are still disabled!\n")
+    LOG_ENABLE();
+    LOG("12 Hello World this time in a new file ( you should not see _2_ in the log or the output )?\n")
+    LOG_SET_TARGET("llama.yetanotherlog.log");
+    LOG("13 Hello World this time in yet new file?\n")
+    LOG_SET_TARGET(LOG_FILENAME_GENERATOR("llama_autonamed", "log"));
+    LOG("14 Hello World in log with generated filename!\n")
+#ifdef _WIN32
+    LOG_TEE("15 Hello msvc TEE without arguments\n")
+    LOG_TEE("16 Hello msvc TEE with (%d)(%s) arguments\n", 1, "test")
+    LOG_TEELN("17 Hello msvc TEELN without arguments\n")
+    LOG_TEELN("18 Hello msvc TEELN with (%d)(%s) arguments\n", 1, "test")
+    LOG("19 Hello msvc LOG without arguments\n")
+    LOG("20 Hello msvc LOG with (%d)(%s) arguments\n", 1, "test")
+    LOGLN("21 Hello msvc LOGLN without arguments\n")
+    LOGLN("22 Hello msvc LOGLN with (%d)(%s) arguments\n", 1, "test")
+#endif
+}
+
+inline bool log_param_single_parse(std::string param)
+{
+    if (std::string("--log-test").compare(param) == 0)
```

Comment:
Same here - why not just write `if (param == "--log-test")`? C++ is smarter than you think.